### PR TITLE
Make test_export training IR compatible

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138517

In this PR, I make test_export to be compatible with training IR. The idea is that when we flip the IR to non-functional training IR, all these tests should be green. The changes involve reading through the test case, and add necessary decomposition etc to make sure the tests pass. For example, if the tests expect to see mutated buffers returned, we need to get them via running run_decomp. 

Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64732360).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64732360).

File: test/export/test_export.py [hunk 1]
```diff
@@ -870,7 +876,7 @@ def forward(self, x):
                 # z = 4
                 return x + y + z + w2
 
-        ep = export(M(), (torch.randn(2, 3),), strict=False)
+        ep = export(M(), (torch.randn(2, 3),), strict=False).run_decompositions({})
```
Comment (avikchaudhuri) [inline test/export/test_export.py:879]: Why these extra `run_decompositions`? I guess I'm missing context on this PR, what does it mean to make these tests "training IR compatible"? Please add a summary to the PR.

Comment (tugsbayasgalan) [reply test/export/test_export.py:879]: The test expects to see mutated buffer as extra output. Only way we get it today is via functiionalizaiton.

File: test/export/test_export.py [hunk 2]
```diff
@@ -1160,47 +1166,6 @@ def forward(self, x):
         )
         check_users_for_graph(ep.graph)
 
-    def test_export_predispatch_custom_ops_warnings(self):
```
Comment (tugsbayasgalan) [inline test/export/test_export.py:1163]: It doesn't matter anymore

File: test/export/test_export.py [hunk 3]
```diff
@@ -3955,7 +3920,7 @@ class Module(torch.nn.Module):
             def forward(self, x):
                 return x.to("cpu")
 
-        ep = export(Module(), (torch.tensor(1, device="cpu"),))
+        ep = export(Module(), (torch.tensor(1, device="cpu"),)).run_decompositions({})
```
Comment (tugsbayasgalan) [inline test/export/test_export.py:3923]: Tests expect to see functional operator.

File: test/export/test_export.py [hunk 4]
```diff
@@ -5545,7 +5461,7 @@ def forward(self, x):
         unflattened = unflatten(ep)
         self.assertTrue(torch.allclose(unflattened(*inps), M2()(*inps)))
 
-    @testing.expectedFailureRetraceability  # Retracing tensor constants results in buffers
+    #@testing.expectedFailureRetraceability  # Retracing tensor constants results in buffers
```
Comment (avikchaudhuri) [inline test/export/test_export.py:5464]: So this goes away, is that good or bad? Just want to make sure that if there's still something to fix, we keep an expected failure.

Comment (tugsbayasgalan) [reply test/export/test_export.py]: The failure was actually not related to tensor constants lol. It was just some graph mismatch.

File: test/export/test_export.py [hunk 5]
```diff
@@ -4257,71 +4236,18 @@ def forward(self, x):
...
-        #     %_assert_async : [num_users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%detach, Fail), kwargs = {})
-        #     return (x,)
-        # The difference comes from the fact that prim has registration for aten.detach while not for aten.detach_.
-        # The diference in retracing comes from the fact that we retrace at pre-dispatch level while the usual flow
-        # traces to post-dispatch.
-        if is_training_ir_test(self._testMethodName):
-            self.assertExpectedInline(
-                str(ep.graph_module.code).strip(),
-                """\
+        ep = export_for_training(M(), (torch.randn(2, 2),)).run_decompositions(
```
Comment (avikchaudhuri) [inline test/export/test_export.py:4239]: looks like you need to remove the xfail from here

**Review by avikchaudhuri — APPROVED** [review]

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64732360).

**Pull request closed without merging**